### PR TITLE
[FP16] Memory efficient optimizer

### PR DESCRIPTION
**Pull request by emilydinan**
**Patch description**
Add a wrapper for a memory efficient FP16 optimizer. Based heavily on the fairseq implementation, which can be found here: https://github.com/pytorch/fairseq/blob/master/fairseq/optim/fp16_optimizer.py. Preliminary tests seems to show this saves about ~6.5 GB when finetuning GPT2-large.

Additionally, added a compatible Adam and Adafactor optimizer (also based on fairseq implementation). 

**Testing steps**
`python -m pytest /private/home/edinan/ParlAI/tests/test_mem_efficient_fp16.py`

### CODE REVIEW

File: parlai/core/torch_agent.py [hunk 1]
```diff
@@ -433,6 +437,13 @@ def add_cmdline_args(cls, argparser):
         agent.add_argument(
             '--fp16', type='bool', default=False, help='Use fp16 computations.'
         )
+        agent.add_argument(
+            '--fp16-impl',
+            type=str,
+            default='apex',
+            choices=['apex', 'mem_efficient'],
```
Comment (EricMichaelSmith) [inline parlai/core/torch_agent.py]: Under what conditions would we still want to use apex? Maybe for the optimizers that this won't work with?

File: parlai/core/torch_agent.py [hunk 2]
```diff
@@ -618,7 +629,12 @@ def __init__(self, opt: Opt, shared=None):
             if not shared and opt['gpu'] != -1:
                 torch.cuda.set_device(opt['gpu'])
         # indicate whether using fp16
-        self.fp16 = self.use_cuda and self.opt.get('fp16', False) and fp16_available()
+        self.fp16 = self.use_cuda and self.opt.get('fp16', False)
+        if self.fp16:
+            # check that the implementation requested is available
+            self.fp16_impl = self.opt.get('fp16_impl', 'apex')
+            if self.fp16_impl == 'apex' and not fp16_apex_available():
+                self.fp16 = False
```
Comment (EricMichaelSmith) [inline parlai/core/torch_agent.py]: Should we print a warning in this case?

File: parlai/core/torch_agent.py [hunk 3]
```diff
@@ -814,11 +833,15 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
...
 
+        # TODO: if self.fp16 and self.fp16_impl == 'mem_efficient'... check that class is ok
         optim_class = self.optim_opts()[opt['optimizer']]
         self.optimizer = optim_class(params, **kwargs)
         if self.fp16:
-            self.optimizer = fp16_optimizer_wrapper(self.optimizer)
-
+            if self.fp16_impl == 'apex':
+                self.optimizer = fp16_optimizer_wrapper(self.optimizer)
+            else:
```
Comment (EricMichaelSmith) [inline parlai/core/torch_agent.py]: nit: `elif self.fp16_impl == 'mem_efficient'` here in case we want to add further choices down the line?

File: parlai/utils/fp16.py [hunk 4]
```diff
@@ -0,0 +1,466 @@
...
+    params = list(params)
+    if len(params) == 1:
+        p = params[0]
+        grad_norm = torch.norm(p)
+        if grad_norm > max_norm > 0:
+            clip_coef = max_norm / (grad_norm + 1e-6)
+            p.mul_(clip_coef)
+        return grad_norm
+    elif max_norm > 0:
+        return torch.nn.utils.clip_grad_norm_(params, max_norm)
```
Comment (EricMichaelSmith) [inline parlai/utils/fp16.py]: hmm - why not use `torch.nn.utils.clip_grad_norm_()` if len(params) == 1?

File: parlai/utils/fp16.py [hunk 5]
```diff
@@ -0,0 +1,466 @@
...
+        self,
+        init_optimizer: torch.optim.Optimizer,  # type: ignore
+        loss_initial_scale: float = 2.0 ** 17,
+        min_loss_scale: float = 1e-4,
+    ):
+        self.optimizer = init_optimizer
+        # TODO: we should probably set a bunch of these args with opt?
+        self.min_loss_scale = min_loss_scale
+        self.scaler = DynamicLossScaler(init_scale=loss_initial_scale)
+        # TODO: add warning here about which optimizers this works with
```
Comment (EricMichaelSmith) [inline parlai/utils/fp16.py]: 👍

File: parlai/core/torch_agent.py (hunk 1 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: Yeah. So I should write more about the tradeoffs inside the class docstring, but basically (1) memory efficinet is not compatible with everything and (2) you want to use apex if memory constraints are not an issue, as there are stability issues with the memory efficient optimizer

File: parlai/core/torch_agent.py (hunk 2 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: oh yeah. we didn't before but we probably should

Comment (emilydinan) [reply parlai/core/torch_agent.py]: actually the warning occurs in the function

**Comment by emilydinan** [comment]
ready for review

File: parlai/core/torch_agent.py [hunk 6]
```diff
@@ -433,6 +439,13 @@ def add_cmdline_args(cls, argparser):
         agent.add_argument(
             '--fp16', type='bool', default=False, help='Use fp16 computations.'
         )
+        agent.add_argument(
+            '--fp16-impl',
+            type=str,
+            default='apex',
+            choices=['apex', 'mem_efficient'],
```
Comment (klshuster) [inline parlai/core/torch_agent.py]: so we technically don't even need apex anymore to run fp16?

File: parlai/core/torch_agent.py [hunk 7]
```diff
@@ -468,6 +481,13 @@ def add_cmdline_args(cls, argparser):
             help='Epsilon value for Adam optimizers. Set to 1e-6 if your '
             'large model has stability issues, but prefer the default.',
         )
+        optim_group.add_argument(
+            '--adafactor-eps',
+            default='1e-30,1e-3',
+            type='floats',
+            help='Epsilon values for adafactor optimizer: regularization'
```
Comment (klshuster) [inline parlai/core/torch_agent.py]: nit: need a space here at the end of the string

File: parlai/core/torch_agent.py [hunk 8]
```diff
@@ -468,6 +481,13 @@ def add_cmdline_args(cls, argparser):
...
+            'constants for square gradienta nd parameter scale respectively',
```
Comment (klshuster) [inline parlai/core/torch_agent.py]: double nit: `for square gradients and`

File: parlai/core/torch_agent.py [hunk 9]
```diff
@@ -468,6 +481,13 @@ def add_cmdline_args(cls, argparser):
             help='Epsilon value for Adam optimizers. Set to 1e-6 if your '
             'large model has stability issues, but prefer the default.',
         )
+        optim_group.add_argument(
+            '--adafactor-eps',
+            default='1e-30,1e-3',
```
Comment (klshuster) [inline parlai/core/torch_agent.py]: are these recommended values? i.e. might be good to have them marked as `recommended=` in the `add_argument` function

File: parlai/core/torch_agent.py [hunk 10]
```diff
@@ -789,10 +814,25 @@ def init_optim(self, params, optim_states=None, saved_optim_type=None):
...
+            kwargs['beta1'] = opt.get('betas', (0.9, 0.999))[0]
+            kwargs['eps'] = opt['adafactor_eps']
+            kwargs['warmup_init'] = opt.get('warmup_updates', -1) > 0
+
+        if opt['optimizer'] in [
+            'adam',
+            'sparseadam',
+            'fused_adam',
+            'adamax',
+            'qhadam',
```
Comment (klshuster) [inline parlai/core/torch_agent.py]: we don't need to set `betas` with the `mem_eff_adam`?

File: parlai/utils/fp16.py [hunk 11]
```diff
@@ -0,0 +1,677 @@
...
+        return torch.sqrt(
+            sum(p.grad.data.norm() ** 2 for p in params if p.grad is not None)
+        )
+
+
+def has_overflow(grad_norm):
+    """
+    Detect inf and NaN in grad_norm.
+    """
+    if grad_norm == float('inf') or grad_norm != grad_norm:
```
Comment (klshuster) [inline parlai/utils/fp16.py]: woah `grad_norm != grad_norm` works for NaN?

File: parlai/utils/fp16.py [hunk 12]
```diff
@@ -0,0 +1,677 @@
...
+
+
+def fp16_apex_available() -> bool:
+    try:
+        import apex.fp16_utils  # noqa: F401
+
+        return True
+    except ImportError:
+        warn_once(
+            'You set --fp16 true, but fp16 is unavailable. To use fp16, please '
```
Comment (klshuster) [inline parlai/utils/fp16.py]: should we modify this warning to indicate that one could also set `--fp16-impl` to `mem_efficient`?

File: parlai/utils/fp16.py [hunk 13]
```diff
@@ -0,0 +1,677 @@
...
+
+
+###################################################
+## Memory Efficient Wrappers
+###################################################
+
+
+class DynamicLossScaler(object):
+    """
+    Shamelessly stolen from Fairseq.
```
Comment (klshuster) [inline parlai/utils/fp16.py]: 😆

Comment (klshuster) [reply parlai/utils/fp16.py]: We might want to change this wording before merging haha

File: parlai/core/torch_agent.py (hunk 6 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: technically. but you probably only want to use mem_efficient in the case where memory is a concern. the memory efficient version can be a bit unstable, and currently only supports a couple of optimizers (adam, adafactor)

File: parlai/core/torch_agent.py (hunk 10 above)
Comment (emilydinan) [reply parlai/core/torch_agent.py]: no, instead we take the first beta only

File: parlai/utils/fp16.py (hunk 11 above)
Comment (emilydinan) [reply parlai/utils/fp16.py]: apparently! this was taken directly from fairseq!

File: parlai/utils/fp16.py (hunk 13 above)
Comment (stephenroller) [reply parlai/utils/fp16.py]:
```
$ ag stolen
docs/source/conf.py
186:    # Stolen shamelessly from Lasagne! Thanks Lasagne!

parlai/utils/distributed.py
146:    # stolen shamelessly from fairseq
```
